File: app/src/main/java/com/robotnec/budget/fragments/AccountsFragment.java
package com.robotnec.budget.fragments;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.View;

import java.util.List;

import butterknife.Bind;
import butterknife.OnClick;
import com.robotnec.budget.R;
import com.robotnec.budget.adapters.AccountsAdapter;
import com.robotnec.budget.core.domain.Account;
import com.robotnec.budget.core.mvp.presenter.AccountsPresenter;
import com.robotnec.budget.core.mvp.view.AccountsView;
import com.robotnec.budget.navigator.AndroidNavigationBundle;

/**
 * @author zak <zak@swingpulse.com>
 */
public class AccountsFragment extends BasePresenterFragment<AccountsPresenter> implements AccountsView, AccountsAdapter.AccountClickListener {

    @Bind(R.id.toolbar)
    Toolbar mToolbar;

    @Bind(R.id.recycler_accounts)
    RecyclerView mRecyclerAccounts;

    private AccountsAdapter mAccountsAdapter;

    public static Fragment newInstance() {
        return new AccountsFragment();
    }

    @Override
    protected int getLayoutId() {
        return R.layout.fragment_accounts;
    }

    @Override
    protected AccountsPresenter createPresenter() {
        return new AccountsPresenter(this);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initToolbarToggle(mToolbar);
        initAccountsList();
    }

    @OnClick(R.id.fab)
    void onFabAddClicked() {
        mPresenter.addOrUpdateAccount(new AndroidNavigationBundle((AppCompatActivity) getActivity()));
    }

    @Override
    public void displayAccounts(List<Account> accounts) {
        mAccountsAdapter.update(accounts);
    }

    private void initAccountsList() {
        mAccountsAdapter = new AccountsAdapter(getContext(), this);
        mRecyclerAccounts.setLayoutManager(new LinearLayoutManager(getContext()));
        mRecyclerAccounts.setHasFixedSize(true);
        mRecyclerAccounts.setAdapter(mAccountsAdapter);
    }

    @Override
    public void onAccountClick(Account account) {
        AndroidNavigationBundle navigationBundle = new AndroidNavigationBundle((AppCompatActivity) getActivity());
        navigationBundle.setSerializableExtra(account);
        mPresenter.addOrUpdateAccount(navigationBundle);
    }
}

File: app/src/main/java/com/robotnec/budget/app/fragments/AccountsFragment.kt
package com.robotnec.budget.app.fragments

import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import android.view.View
import com.robotnec.budget.R
import com.robotnec.budget.app.adapters.AccountsAdapter
import com.robotnec.budget.app.navigator.AndroidNavigationBundle
import com.robotnec.budget.core.domain.Account
import com.robotnec.budget.core.mvp.presenter.AccountsPresenter
import com.robotnec.budget.core.mvp.view.AccountsView
import kotlinx.android.synthetic.main.fragment_accounts.fab
import kotlinx.android.synthetic.main.fragment_accounts.recyclerAccounts
import kotlinx.android.synthetic.main.fragment_accounts.toolbar

/**
 * @author zak zak@swingpulse.com>
 */
class AccountsFragment : BasePresenterFragment<AccountsPresenter>(), AccountsView, AccountsAdapter.AccountClickListener {

    companion object {
        fun newInstance(): Fragment {
            return AccountsFragment()
        }
    }

    private lateinit var accountsAdapter: AccountsAdapter

    override val layoutId: Int
        get() = R.layout.fragment_accounts

    override fun createPresenter(): AccountsPresenter {
        return AccountsPresenter(this)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initToolbarToggle(toolbar)
        initAccountsList()
        fab.setOnClickListener {
            presenter.addOrUpdateAccount(AndroidNavigationBundle(activity as AppCompatActivity, null))
        }
    }

    override fun displayAccounts(accounts: List<Account>) {
        accountsAdapter.update(accounts)
    }

    private fun initAccountsList() {
        context?.let {
            accountsAdapter = AccountsAdapter(it, this)
            recyclerAccounts.layoutManager = LinearLayoutManager(context)
            recyclerAccounts.adapter = accountsAdapter
        }
    }

    override fun onAccountClick(account: Account) {
        val navigationBundle = AndroidNavigationBundle(activity as AppCompatActivity, account)
        presenter.addOrUpdateAccount(navigationBundle)
    }
}
